pkg/registry/core/service/storage/storage.go

```diff
@@ -99,3 +125,102 @@ func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.Updat
 	// subresources should never allow create on update.
 	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation, false, options)
 }
+
+// defaults fields that were not previously set on read. becomes an
+// essential part of upgrading a service
+func (r *GenericREST) defaultServiceOnRead(obj runtime.Object) error {
+	service, ok := obj.(*api.Service)
+	if ok {
+		return r.defaultSvcOnRead(service)
+	}
+
+	serviceList, ok := obj.(*api.ServiceList)
+	if ok {
+		return r.defaultServiceList(serviceList)
+	}
+
+	// this was not an object we can default
+	return nil
+}
+
+// defaults a service list
+func (r *GenericREST) defaultServiceList(serviceList *api.ServiceList) error {
+	if serviceList == nil {
+		return nil
+	}
+
+	for i, svc := range serviceList.Items {
+		err := r.defaultSvcOnRead(&svc)
+		if err != nil {
+			return err
+		}
+		serviceList.Items[i] = svc
+	}
+
+	return nil
+}
+
+// defaults a single service
+func (r *GenericREST) defaultSvcOnRead(service *api.Service) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		return nil
+	}
+
+	if service == nil {
+		return nil
+	}
+
+	if len(service.Spec.IPFamilies) > 0 {
+		return nil // already defaulted
+	}
+
+	requireDualStack := api.RequireDualStack
+	singleStack := api.SingleStack
+
+	// headless services
+	if len(service.Spec.ClusterIPs) == 1 && service.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		service.Spec.IPFamilies = []api.IPFamily{*r.primaryIPFamily}
+
+		// headless+selectorless
+		// headless+selectorless takes both families
+		if len(service.Spec.Selector) == 0 {
+			service.Spec.IPFamilyPolicy = &requireDualStack
```

Comment:
As above, the KEP currently says `PreferDualStack`.

It also seems like this should share code with the other defaulting code...

Reply:
This is the correct behavior. I will update the KEP.